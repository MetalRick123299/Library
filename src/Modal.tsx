import clsx from 'clsx';
import React, { Dispatch, SetStateAction, useState } from 'react';
import { XCircleIcon } from '@heroicons/react/20/solid';
import { IBookItem } from './BookItem';

interface ModalProps {
  isModal: boolean;
  setIsModal: Dispatch<SetStateAction<boolean>>;
  setBookList: Dispatch<SetStateAction<IBookItem[]>>;
  bookList: IBookItem[];
}

const emptyForm: IBookItem = {
  title: '',
  author: '',
  pagesRead: 0,
  totalPages: 0,
};

export default function Modal({
  isModal,
  setIsModal,
  setBookList,
  bookList,
}: ModalProps) {
  const [formInputs, setFormInputs] = useState<IBookItem>(emptyForm);

  const handleChange = (e: React.FormEvent<HTMLInputElement>) => {
    const currName = e.currentTarget.name.replace(/\s/g, '');
    let currValue: string | number = e.currentTarget.value;

    if (
      currValue !== '' &&
      (currName === 'totalPages' || currName === 'pagesRead')
    ) {
      currValue = parseInt(e.currentTarget.value, 10);
    }
    setFormInputs((prev) => ({ ...prev, [currName]: currValue }));
  };

  const handleSubmit = (e: React.FormEvent<HTMLFormElement>) => {
    const { title, pagesRead, totalPages } = formInputs;
    e.preventDefault();

    if (bookList.findIndex((ele) => ele.title === title) !== -1) return;
    if (pagesRead > totalPages) return;

    setBookList((prev) => [...prev, formInputs]);
    setIsModal(false);
    setFormInputs(emptyForm);
  };

  return (
    <div
      className={clsx(
        'fixed flex flex-col justify-center items-center h-screen w-screen bg-black/60 z-10 text-xl transition-all',
        isModal ? 'scale-100' : 'scale-0'
      )}
      aria-hidden={isModal ? 'false' : 'true'}
    >
      <form
        action=""
        className="flex flex-col items-center gap-3 bg-primary-bg p-5 rounded-xl border-primary-nav border-8"
        aria-label="Add Book Form"
        onSubmit={handleSubmit}
      >
        <label className="">
          Search Book
          <input
            className="bg-primary-item block rounded-md px-2 py-1 mt-1"
            type="text"
            name="search Book"
          />
        </label>

        <div className="text-2xl -mb-3 font-semibold">OR</div>

        <label className="">
          Title
          <input
            className="bg-primary-item block rounded-md px-2 py-1 mt-1"
            type="text"
            name="title"
            value={formInputs.title}
            onChange={handleChange}
            required
          />
        </label>

        <label className="">
          Author
          <input
            className="bg-primary-item block rounded-md px-2 py-1 mt-1"
            type="text"
            name="author"
            value={formInputs.author}
            onChange={handleChange}
            required
          />
        </label>

        <label className="">
          Pages Read
          <input
            type="number"
            className="bg-primary-item block rounded-md px-2 py-1 mt-1"
            name="pages Read"
            value={formInputs.pagesRead}
            onChange={handleChange}
            required
          />
        </label>

        <label className="">
          Total Pages
          <input
            type="number"
            className="bg-primary-item block rounded-md px-2 py-1 mt-1"
            name="total Pages"
            value={formInputs.totalPages}
            onChange={handleChange}
            required
          />
        </label>

        <div className="flex items-start gap-3 bg-primary-item p-3 rounded-xl">
          <button type="submit" className="" aria-label="Add Book Submit">
            Add Book
          </button>
          <XCircleIcon
            className="btn inline !h-7"
            role="button"
            aria-hidden="false"
            aria-label="Close Form Button"
            onClick={() => setIsModal(false)}
          />
        </div>
      </form>
    </div>
  );
}
