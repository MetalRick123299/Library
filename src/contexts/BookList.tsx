/* eslint-disable import/prefer-default-export */
import React, {
  createContext,
  useMemo,
  Dispatch,
  SetStateAction,
  ReactNode,
  useState,
  useContext,
} from 'react';

export type IBookItem = {
  title: string;
  author: string;
  pagesRead: number;
  totalPages: number;
};

const book: IBookItem = {
  title: 'Harry Potter',
  author: 'Maybe JK',
  pagesRead: 256,
  totalPages: 586,
};

export const books: IBookItem[] = [
  { ...book, title: 'Book Title 1' },
  { ...book, pagesRead: 0, title: 'Book Title 2' },
  { ...book, title: 'Book Title 3' },
  { ...book, pagesRead: book.totalPages, title: 'Book Title 4' },
  { ...book, title: 'Book Title 5' },
];

interface IBookListContext {
  bookList: IBookItem[];
  setBookList: Dispatch<SetStateAction<IBookItem[]>>;
}
export const BookListContext = createContext<IBookListContext>(null!);
BookListContext.displayName = 'BookListContext';

export function BookListProvider({ children }: { children: ReactNode }) {
  const [bookList, setBookList] = useState<IBookItem[]>(books);

  return (
    // eslint-disable-next-line react/jsx-no-constructed-context-values
    <BookListContext.Provider value={{ bookList, setBookList }}>
      {children}
    </BookListContext.Provider>
  );
}
