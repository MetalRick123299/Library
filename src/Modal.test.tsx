import { render, screen, fireEvent } from '@testing-library/react';
import Modal, { emptyForm } from './Modal';
import { IBookItem, BookListProvider } from './contexts/BookList';

describe('Modal Tests', () => {
  const book: IBookItem = {
    title: 'Harry Potter',
    author: 'J.K Rowling',
    pagesRead: 45,
    totalPages: 586,
  };
  const mockFn = vi.fn();

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  let modal: any;
  beforeEach(() => {
    modal = render(
      <BookListProvider>
        <Modal isModal setIsModal={mockFn} initForm={emptyForm} />
      </BookListProvider>
    );
  });

  it('Modal Structure', () => {
    const titleInput = screen.getByRole('textbox', { name: /title/i });
    const authorInput = screen.getByRole('textbox', { name: /author/i });
    const pagesReadInput = screen.getByRole('spinbutton', {
      name: /Pages Read/i,
    });
    const totalPagesInput = screen.getByRole('spinbutton', {
      name: /Total Pages/i,
    });

    fireEvent.change(titleInput, { target: { value: book.title } });
    fireEvent.change(authorInput, { target: { value: book.author } });
    fireEvent.change(pagesReadInput, { target: { value: book.pagesRead } });
    fireEvent.change(totalPagesInput, { target: { value: book.totalPages } });
  });
  it('Modal Snapshot', () => {
    expect(modal).toMatchSnapshot();
  });

  it('No Pages Read more than Total Pages', () => {
    const titleInput = screen.getByRole('textbox', { name: /title/i });
    const authorInput = screen.getByRole('textbox', { name: /author/i });
    const pagesReadInput = screen.getByRole('spinbutton', {
      name: /Pages Read/i,
    });
    const totalPagesInput = screen.getByRole('spinbutton', {
      name: /Total Pages/i,
    });

    fireEvent.change(titleInput, { target: { value: book.title } });
    fireEvent.change(authorInput, { target: { value: book.author } });
    fireEvent.change(pagesReadInput, { target: { value: book.totalPages } });
    fireEvent.change(totalPagesInput, { target: { value: book.pagesRead } });

    let addBookSubmit = screen.getByRole('button', {
      name: 'Add Book Submit',
    });
    fireEvent.click(addBookSubmit);

    addBookSubmit = screen.getByRole('button', {
      name: 'Add Book Submit',
    });

    expect(addBookSubmit).not.toBe(null);
  });
});
