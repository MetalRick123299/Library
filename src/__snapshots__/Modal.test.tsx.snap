// Vitest Snapshot v1

exports[`Modal Tests > Modal Snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        aria-hidden="false"
        class="fixed flex flex-col justify-center items-center h-screen w-screen bg-black/60 z-10 text-xl transition-all scale-100"
      >
        <form
          action=""
          aria-label="Edit Book Form"
          class="flex flex-col items-center gap-3 bg-primary-bg p-5 rounded-xl border-primary-nav border-8"
        >
          <label
            class=""
          >
            Search Book
            <input
              class="bg-primary-item block rounded-md px-2 py-1 mt-1"
              name="search Book"
              type="text"
            />
          </label>
          <div
            class="text-2xl -mb-3 font-semibold"
          >
            OR
          </div>
          <label
            class=""
          >
            Title
            <input
              class="bg-primary-item block rounded-md px-2 py-1 mt-1"
              name="title"
              required=""
              type="text"
              value="Harry Potter"
            />
          </label>
          <label
            class=""
          >
            Author
            <input
              class="bg-primary-item block rounded-md px-2 py-1 mt-1"
              name="author"
              required=""
              type="text"
              value="J.K Rowling"
            />
          </label>
          <label
            class=""
          >
            Pages Read
            <input
              class="bg-primary-item block rounded-md px-2 py-1 mt-1"
              name="pages Read"
              required=""
              type="number"
              value="45"
            />
          </label>
          <label
            class=""
          >
            Total Pages
            <input
              class="bg-primary-item block rounded-md px-2 py-1 mt-1"
              name="total Pages"
              required=""
              type="number"
              value="586"
            />
          </label>
          <div
            class="flex items-start gap-3 bg-primary-item p-3 rounded-xl"
          >
            <button
              aria-label="Add Book Submit"
              class=""
              type="submit"
            >
              Edit Form
            </button>
            <svg
              aria-hidden="false"
              aria-label="Close Form Button"
              class="btn inline !h-7"
              fill="currentColor"
              role="button"
              viewBox="0 0 20 20"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                clip-rule="evenodd"
                d="M10 18a8 8 0 100-16 8 8 0 000 16zM8.28 7.22a.75.75 0 00-1.06 1.06L8.94 10l-1.72 1.72a.75.75 0 101.06 1.06L10 11.06l1.72 1.72a.75.75 0 101.06-1.06L11.06 10l1.72-1.72a.75.75 0 00-1.06-1.06L10 8.94 8.28 7.22z"
                fill-rule="evenodd"
              />
            </svg>
          </div>
        </form>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      aria-hidden="false"
      class="fixed flex flex-col justify-center items-center h-screen w-screen bg-black/60 z-10 text-xl transition-all scale-100"
    >
      <form
        action=""
        aria-label="Edit Book Form"
        class="flex flex-col items-center gap-3 bg-primary-bg p-5 rounded-xl border-primary-nav border-8"
      >
        <label
          class=""
        >
          Search Book
          <input
            class="bg-primary-item block rounded-md px-2 py-1 mt-1"
            name="search Book"
            type="text"
          />
        </label>
        <div
          class="text-2xl -mb-3 font-semibold"
        >
          OR
        </div>
        <label
          class=""
        >
          Title
          <input
            class="bg-primary-item block rounded-md px-2 py-1 mt-1"
            name="title"
            required=""
            type="text"
            value="Harry Potter"
          />
        </label>
        <label
          class=""
        >
          Author
          <input
            class="bg-primary-item block rounded-md px-2 py-1 mt-1"
            name="author"
            required=""
            type="text"
            value="J.K Rowling"
          />
        </label>
        <label
          class=""
        >
          Pages Read
          <input
            class="bg-primary-item block rounded-md px-2 py-1 mt-1"
            name="pages Read"
            required=""
            type="number"
            value="45"
          />
        </label>
        <label
          class=""
        >
          Total Pages
          <input
            class="bg-primary-item block rounded-md px-2 py-1 mt-1"
            name="total Pages"
            required=""
            type="number"
            value="586"
          />
        </label>
        <div
          class="flex items-start gap-3 bg-primary-item p-3 rounded-xl"
        >
          <button
            aria-label="Add Book Submit"
            class=""
            type="submit"
          >
            Edit Form
          </button>
          <svg
            aria-hidden="false"
            aria-label="Close Form Button"
            class="btn inline !h-7"
            fill="currentColor"
            role="button"
            viewBox="0 0 20 20"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              clip-rule="evenodd"
              d="M10 18a8 8 0 100-16 8 8 0 000 16zM8.28 7.22a.75.75 0 00-1.06 1.06L8.94 10l-1.72 1.72a.75.75 0 101.06 1.06L10 11.06l1.72 1.72a.75.75 0 101.06-1.06L11.06 10l1.72-1.72a.75.75 0 00-1.06-1.06L10 8.94 8.28 7.22z"
              fill-rule="evenodd"
            />
          </svg>
        </div>
      </form>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
